File: android/app/src/main/java/com/clovis/app/connecteurs/ConnecteursPlugin.kt
// Cree le 25/08/2026, Bourama : Lot 3B Partie 3 mobile (fusion Capacitor).
// Porte depuis clovis-mobile/android-legacy-natif (Lot 5, ConnecteursScreen.kt
// + OAuthCallbackActivity.kt + ClovisApiClient methodes Notion) : meme flow
// Custom Tabs + retour sur clovismobile://oauth-callback, ecran retire.
//
// Cote JS :
//   const Connecteurs = registerPlugin<any>('Connecteurs');
//   await Connecteurs.statutNotion();                    // { connecte: bool }
//   await Connecteurs.demarrerConnexionNotion();          // ouvre Custom Tabs (Notion)
//   await Connecteurs.rechercherNotion({ requete });
//   Connecteurs.addListener('retourOAuth', ({ code, state }) => { ... }) // apres finaliserConnexionNotion cote backend
// L'app doit appeler finaliserConnexionNotion() elle-meme (pas ce plugin) une
// fois code/state recus, comme le faisait ConnecteursScreen.kt, ce plugin
// se contente de relayer l'evenement OAuth, pas de decider la logique.
package com.clovis.app.connecteurs

import android.content.Intent
import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import com.clovis.app.pont.ClovisApiClient
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

@CapacitorPlugin(name = "Connecteurs")
class ConnecteursPlugin : Plugin() {

    private val client by lazy { ClovisApiClient(context) }

    override fun load() {
        super.load()
        // S'enregistre pour recevoir l'evenement publie par OAuthCallbackActivity
        // (meme mecanisme RetourOAuth que la version legacy, garde tel quel).
        CoroutineScope(Dispatchers.Main).launch {
            RetourOAuth.evenements.collect { (code, state) ->
                notifyListeners("retourOAuth", JSObject().put("code", code).put("state", state))
            }
        }
    }

    @PluginMethod
    fun statutNotion(call: PluginCall) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val statut = client.statutNotion()
                call.resolve(JSObject().put("connecte", statut.connecte))
            } catch (e: Exception) {
                call.reject("Echec de la lecture du statut Notion.", e)
            }
        }
    }

    @PluginMethod
    fun demarrerConnexionNotion(call: PluginCall) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val reponse = client.demarrerConnexionNotion()
                CoroutineScope(Dispatchers.Main).launch {
                    val customTabsIntent = CustomTabsIntent.Builder().build()
                    customTabsIntent.launchUrl(context, Uri.parse(reponse.url_autorisation))
                    call.resolve()
                }
            } catch (e: Exception) {
                call.reject("Echec du demarrage de la connexion Notion.", e)
            }
        }
    }

    @PluginMethod
    fun finaliserConnexionNotion(call: PluginCall) {
        val code = call.getString("code")
        val state = call.getString("state")
        if (code == null || state == null) {
            call.reject("Parametres 'code' et 'state' requis.")
            return
        }
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val reponse = client.finaliserConnexionNotion(code, state)
                call.resolve(JSObject().put("connecte", reponse.connecte).put("espace", reponse.espace))
            } catch (e: Exception) {
                call.reject("Echec de la finalisation de la connexion Notion.", e)
            }
        }
    }

    @PluginMethod
    fun rechercherNotion(call: PluginCall) {
        val requete = call.getString("requete")
        if (requete == null) {
            call.reject("Parametre 'requete' manquant.")
            return
        }
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val reponse = client.rechercherNotion(requete)
                val tableau = com.getcapacitor.JSArray()
                reponse.resultats.forEach {
                    tableau.put(JSObject().put("id", it.id).put("type", it.type).put("url", it.url))
                }
                call.resolve(JSObject().put("resultats", tableau))
            } catch (e: Exception) {
                call.reject("Echec de la recherche Notion.", e)
            }
        }
    }
}

File: ios/App/App/connecteurs/ConnecteursPlugin.swift
// Cree le 25/08/2026, Bourama : Lot 3B Partie 3 mobile (fusion Capacitor), iOS.
// Porte depuis clovis-mobile/ios-legacy-natif (Lot 5, ConnecteursScreen.swift) :
// meme mecanisme ASWebAuthenticationSession (callbackURLScheme="clovismobile"),
// qui n'a PAS besoin d'etre declare dans Info.plist (CFBundleURLTypes) :
// contrairement a l'equivalent Android (Custom Tabs + intent-filter dans
// AndroidManifest.xml), la session intercepte elle-meme sa propre navigation
// avant que l'OS n'essaie de la router.
//
// Cote JS (identique Android en surface, mais demarrerConnexionNotion() sur
// iOS est BLOQUANT : il gere lui-meme code+state en interne et resout
// directement le statut, pas besoin d'ecouter un evenement retourOAuth
// separe comme sur Android) :
//   const Connecteurs = registerPlugin<any>('Connecteurs');
//   await Connecteurs.statutNotion();
//   await Connecteurs.demarrerConnexionNotion();  // ouvre + attend + finalise, tout-en-un
//   await Connecteurs.rechercherNotion({ requete });
import Foundation
import Capacitor
import AuthenticationServices
import UIKit

@objc(ConnecteursPlugin)
public class ConnecteursPlugin: CAPPlugin, CAPBridgedPlugin, ASWebAuthenticationPresentationContextProviding {
    public let identifier = "ConnecteursPlugin"
    public let jsName = "Connecteurs"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "statutNotion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "demarrerConnexionNotion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "rechercherNotion", returnType: CAPPluginReturnPromise)
    ]

    // Garde une reference forte le temps de la session, sinon ARC la libere
    // avant la fin du flux (meme piege que SessionConnexionNotion legacy,
    // qui s'en sortait via @StateObject -- pas disponible ici).
    private var sessionEnCours: ASWebAuthenticationSession?

    public func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        DispatchQueue.main.sync {
            UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .first?.windows.first { $0.isKeyWindow } ?? ASPresentationAnchor()
        }
    }

    @objc func statutNotion(_ call: CAPPluginCall) {
        Task {
            do {
                let statut = try await ClovisApiClient.statutNotion()
                call.resolve(["connecte": statut.connecte])
            } catch {
                call.reject("Echec de la lecture du statut Notion.", nil, error)
            }
        }
    }

    @objc func demarrerConnexionNotion(_ call: CAPPluginCall) {
        Task {
            do {
                let reponse = try await ClovisApiClient.demarrerConnexionNotion()
                guard let url = URL(string: reponse.url_autorisation) else {
                    call.reject("URL d'autorisation Notion invalide.")
                    return
                }
                let (code, state) = try await lancerSessionOAuth(url: url)
                let resultat = try await ClovisApiClient.finaliserConnexionNotion(code: code, state: state)
                call.resolve(["connecte": resultat.connecte, "espace": resultat.espace as Any])
            } catch {
                call.reject("Connexion Notion echouee.", nil, error)
            }
        }
    }

    @MainActor
    private func lancerSessionOAuth(url: URL) async throws -> (code: String, state: String) {
        try await withCheckedThrowingContinuation { continuation in
            let session = ASWebAuthenticationSession(url: url, callbackURLScheme: "clovismobile") { callbackURL, error in
                if let error {
                    continuation.resume(throwing: error)
                    return
                }
                guard
                    let callbackURL,
                    let composants = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false),
                    let code = composants.queryItems?.first(where: { $0.name == "code" })?.value,
                    let state = composants.queryItems?.first(where: { $0.name == "state" })?.value
                else {
                    continuation.resume(throwing: NSError(
                        domain: "ConnecteursPlugin", code: 1,
                        userInfo: [NSLocalizedDescriptionKey: "Reponse de connexion incomplete."]
                    ))
                    return
                }
                continuation.resume(returning: (code, state))
            }
            session.presentationContextProvider = self
            session.prefersEphemeralWebBrowserSession = true
            self.sessionEnCours = session
            session.start()
        }
    }

    @objc func rechercherNotion(_ call: CAPPluginCall) {
        guard let requete = call.getString("requete") else {
            call.reject("Parametre 'requete' manquant.")
            return
        }
        Task {
            do {
                let reponse = try await ClovisApiClient.rechercherNotion(requete)
                let resultats = reponse.resultats.map {
                    ["id": $0.id, "type": $0.type, "url": $0.url as Any] as [String: Any]
                }
                call.resolve(["resultats": resultats])
            } catch {
                call.reject("Echec de la recherche Notion.", nil, error)
            }
        }
    }
}
